extensions/message-edit.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Message editing and deletion
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "James Wheare"
+    email: "james@irccloud.com"
+    period: "2020"
+contributors:
+  -
+    name: "jesopo"
+    email: "jess@jesopo.uk"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the unprefixed `message-edit` or `message-delete` capability names. Instead, implementations SHOULD use the `draft/message-edit` and `draft/message-delete` capability names to be interoperable with other software implementing a compatible work-in-progress version.
+
+The final version of the specification will use an unprefixed capability name.
+
+
+## Introduction
+
+This specification enables messages to be edited and deleted. Use cases include typo correction, retracting accidentally sent messages, and moderation, amongst others. These are cosmetic use cases and do not provide any operational security guarantees.
+
+## Architecture
+
+### Dependencies
+
+Clients wishing to use these capabilities MUST negotiate the [`message-tags`](../extensions/message-tags.html) capability with the server. Additionally, this capability relies on messages being sent with the [`msgid`](../extensions/message-ids.html) tag. Clients SHOULD negotiate the [`echo-message`](../extensions/echo-message-3.2.html) and [`labeled-response`](../extensions/labeled-response.html) capabilities in order to receive message IDs for their own messages, to allow them to be edited and deleted.
+
+### Capabilities
+
+This specification adds the `draft/message-edit` and `draft/message-delete` capabilities.
+
+Implementations that negotiate these capabilities indicate that they are capable of handling the respective commands and message tags described below.
+
+### Editing messages
+
+To edit a message, a client MUST negotiate the `draft/message-edit` capability and send an `EDIT` command to a target nickname or channel. The command is defined as follows:
+
+    @<tags> EDIT <target>
+
+And uses the following tags:
+
+* `draft/target-msgid=<msgid>` to indicate the [`msgid`] of the message to be edited
+* `draft/edit-text=<new-text>` to indicate the new text of the message
+
+If the client is authorised to edit the message, the server MUST forward this `EDIT` to the target recipients with an appropriate source.
```

Comment:
Why `MUST`? Nothing bad will happen if it doesn't, and there are many reasons for a server not to forward it (eg. netsplit).

And `source` is not defined anywhere (it's only used informally in 1459, and not at all in this context in 2812); what about `prefix` or `prefix as source` instead?

Reply:
This is the same value of MUST as the contract that if I send a well-formed PRIVMSG and have all the relevant permissions, the server sends it to the target. What's the issue?

I'll change source to prefix.

Reply:
> if I send a well-formed PRIVMSG and have all the relevant permissions, the server sends it to the target

Where is it specified as a MUST?

> What's the issue?

Like I said, sending it is not an absolute requirement of the spec, and there are valid reasons not to send it in particular circumstances. So it should be a SHOULD, not a MUST.

Reply:
> Where is it specified as a MUST?

Dunno if the RFCs are the best example to go looking for well-specified robust language, but the expectation is certainly that if the server accepts the PRIVMSG and doesn't return an error, it passes it along (disregarding stuff like SHUN that I don't really think we need to worry about in spec language).

If there's a valid reason not to send the message along, there should be an error message.

Reply:
Also, for a situation where the server is literally disconnected, obviously it won't be sent along, we don't have to make every spec wooly and toothless to account for network issues.

Reply:
Alternatively, to borrow wording from the `message-tags` spec (`The TAGMSG tag-only message` section): "If the client is authorised to edit the message this message MUST be delivered to targets in the same way as PRIVMSG and NOTICE messages."

Reply:
Yeah, I mean that is definitely the intention, works for me. Can probably drop the "and NOTICE" part as it doesn't add anything mechanics wise. Would that satisfy your concern?

Reply:
yes, sure